src/librustc/ty/maps/plumbing.rs

```diff
@@ -79,32 +81,43 @@ pub(super) enum TryGetLock<'a, 'tcx: 'a, T, D: QueryDescription<'tcx> + 'a> {
 }
 
 impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
-    pub(super) fn report_cycle(self, CycleError { span, cycle: stack }: CycleError)
+    pub(super) fn report_cycle(self, CycleError { usage, cycle: stack }: CycleError<'gcx>)
         -> DiagnosticBuilder<'a>
     {
         assert!(!stack.is_empty());
 
+        let fix_span = |span: Span, query: &Query<'gcx>| {
+            self.sess.codemap().def_span(query.default_span(self, span))
+        };
+
         // Disable naming impls with types in this path, since that
         // sometimes cycles itself, leading to extra cycle errors.
         // (And cycle errors around impls tend to occur during the
         // collect/coherence phases anyhow.)
         item_path::with_forced_impl_filename_line(|| {
-            let span = self.sess.codemap().def_span(span);
-            let mut err =
-                struct_span_err!(self.sess, span, E0391,
-                                 "cyclic dependency detected");
-            err.span_label(span, "cyclic reference");
-
-            err.span_note(self.sess.codemap().def_span(stack[0].span),
-                          &format!("the cycle begins when {}...", stack[0].query.describe(self)));
-
-            for &QueryInfo { span, ref query, .. } in &stack[1..] {
-                err.span_note(self.sess.codemap().def_span(span),
-                              &format!("...which then requires {}...", query.describe(self)));
+            let mut err = struct_err!(self.sess,
```

Comment:
You could just use the span of `stack[0]`, since that's the one mentioned in the error message

Reply:
I don't want to print the same span twice. The message is quite noisy already.

Reply:
You can start the iteration at i=1, the it's not printed twice

Reply:
We should _avoid_ diagnostics without a main span. If verbosity is the problem, lets rewrite the diagnostic so that the redundant span note is turned into a note for the main span. This might require a complete rewrite of the text to make sense, specially in cases where a chain of causality is tried to be shown and explained.

Reply:
I think I can get away with just shifting the spans around without changing any messages. The text sufficiently unclear to allow that...